File: packages/igloo-ts/src/widgets/button.ts
import { Button as WitButton } from 'iced:app/button@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Padding } from 'iced:app/padding@0.1.0';
import { Element, toElement, type IntoElement, type ElementLike } from '../element.js';
import { pushFixed } from '../callbacks.js';

/**
 * Builder for creating Button widgets.
 *
 * @example
 * ```typescript
 * const button = Button.new(Text.new('Click me'))
 *   .onPress(() => ({ type: 'clicked' }))
 *   .padding(Padding.all(10));
 *
 * Column.new().push(button);
 * ```
 *
 * @typeParam Msg - The message type this button emits; inferred from `onPress`.
 */
export class Button<Msg = never> implements IntoElement<Msg> {
  private raw: WitButton;

  private constructor(content: ElementLike<never>) {
    this.raw = new WitButton(toElement(content).inner);
  }

  /** Create a new Button builder with the given content */
  static new(content: ElementLike<never>): Button<never> {
    return new Button(content);
  }

  /** Set the message to emit when the button is pressed */
  onPress<const M>(msg: () => M): Button<Msg | M> {
    this.raw.onPress(pushFixed(msg()));
    // The runtime object is unchanged; only the static type advances to carry M.
    return this as unknown as Button<Msg | M>;
  }

  /** Set the button width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the button height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set the button padding */
  padding(padding: Padding): this {
    this.raw.padding(padding);
    return this;
  }

  /** Enable or disable clipping of content */
  clip(clip: boolean = true): this {
    this.raw.clip(clip);
    return this;
  }

  /** Convert to Element (implements IntoElement) */
  intoElement(): Element<Msg> {
    return new Element(WitButton.intoElement(this.raw));
  }
}

File: packages/igloo-py/src/igloo_py/widgets/button.py
"""Button widget builder."""

from collections.abc import Callable
from typing import TYPE_CHECKING, TypeVar

from wit_world.imports.button import Button as WitButton

from ..callbacks import push_fixed
from ..element import Element, ElementLike, IntoElement, to_element

if TYPE_CHECKING:
    from ..types.length import WitLength
    from ..types.padding import WitPadding

Msg = TypeVar("Msg")


class Button(IntoElement):
    """
    Builder for creating Button widgets.

    Example:
        # Widgets can be passed directly - no into_element() needed
        button = Button.new(Text.new("Click me"))
            .on_press(lambda: {"type": "clicked"})
            .padding(Padding.all(10))

        # Use in a Column directly
        Column.new().push(button)
    """

    def __init__(self, content: ElementLike) -> None:
        self._raw = WitButton(to_element(content).inner)

    @classmethod
    def new(cls, content: ElementLike) -> "Button":
        """Create a new Button builder with the given content."""
        return cls(content)

    def on_press(self, message: Callable[[], Msg]) -> "Button":
        """Set the message to emit when the button is pressed."""
        self._raw.on_press(push_fixed(message()))
        return self

    def width(self, width: "WitLength") -> "Button":
        """Set the button width."""
        self._raw.width(width)
        return self

    def height(self, height: "WitLength") -> "Button":
        """Set the button height."""
        self._raw.height(height)
        return self

    def padding(self, padding: "WitPadding") -> "Button":
        """Set the button padding."""
        self._raw.padding(padding)
        return self

    def clip(self, clip: bool = True) -> "Button":
        """Enable or disable clipping of content."""
        self._raw.clip(clip)
        return self

    def into_element(self) -> Element:
        """Convert to Element (implements IntoElement)."""
        return Element(WitButton.into_element(self._raw))
